FirebaseFirestoreUI/FUISnapshotArrayDiff.m

```diff
@@ -377,6 +377,131 @@ - (instancetype)initWithInitialArray:(NSArray<FIRDocumentSnapshot *> *)initial
   return self;
 }
 
+- (void)buildFastDiffs {
+  NSMutableDictionary<NSString *, NSNumber *> *oldIndexes =
+      [NSMutableDictionary dictionaryWithCapacity:_initial.count];
+  NSMutableDictionary<NSString *, NSNumber *> *newIndexes =
+      [NSMutableDictionary dictionaryWithCapacity:_result.count];
+
+  NSArray<FIRDocumentSnapshot *> *initial = _initial;
+  NSArray<FIRDocumentSnapshot *> *result = _result;
+
+  for (NSInteger i = 0; i < initial.count; i++) {
+    oldIndexes[initial[i].documentID] = @(i);
+  }
+  for (NSInteger i = 0; i < result.count; i++) {
+    newIndexes[result[i].documentID] = @(i);
+  }
+
+  // This set is used to look up the index of deletions later on.
+  NSMutableOrderedSet<NSNumber *> *deletedIndexes = [NSMutableOrderedSet orderedSet];
+  NSMutableOrderedSet<FIRDocumentSnapshot *> *deletedObjects = [NSMutableOrderedSet orderedSet];
+
+  NSMutableArray<NSNumber *> *insertedIndexes = [NSMutableArray array];
+  NSMutableArray<FIRDocumentSnapshot *> *insertedObjects = [NSMutableArray array];
+
+  NSMutableArray<NSNumber *> *changedIndexes = [NSMutableArray array];
+  NSMutableArray<FIRDocumentSnapshot *> *changedObjects = [NSMutableArray array];
+
+  NSMutableArray<NSNumber *> *movedInitialIndexes = [NSMutableArray array];
+  NSMutableArray<NSNumber *> *movedResultIndexes = [NSMutableArray array];
+  NSMutableArray<FIRDocumentSnapshot *> *movedObjects = [NSMutableArray array];
+
+  // Loop through the collections, building diff information as we go.
+  // This is simpler than some other diff agorithms because documentIDs are guaranteed
+  // to be unique, so we can't have multiple versions of the same document in a
+  // collection. Here's the observations we're making:
+  //
+  // 1. If the document is unchanged and its start and end indexes are the same, ignore the
+  //    document.
+  // 2. If the document is different in the initial and result arrays but the documentID
+  //    is the same, then the document was changed (and should be reloaded).
+  // 3. If the document did not exist in the initial array, this corresponds to an add.
+  // 4. If the document did not exist in the result array, this corresponds to a delete.
+  // 5. If an add and delete occurred at the same index, this corresponds to a reload.
```

Comment:
For clarity, would it be possible to use the same terminology here as for the actual output arrays. I am assuming that "a reload" corresponds to a "change" (i.e. changedObjects and changedIndexes)?

Reply:
Will do :)